File: crates/folio-qt/src/transaction_table.rs
use crate::{RUNTIME, db_pool};
use folio_lib::service::{Account, Expense, fetch_expenses};
use qmetaobject::{QAbstractTableModel, QObject, USER_ROLE, qt_base_class, qt_method};
use qttypes::{QByteArray, QModelIndex, QString, QVariant};
use sqlx::types::chrono::NaiveDate;
use std::collections::HashMap;

#[derive(QObject, Default)]
pub struct TransactionTableModel {
    base: qt_base_class!(trait QAbstractTableModel),
    load_expenses: qt_method!(fn(&mut self)),
    add_expense: qt_method!(fn(&mut self)),
    expenses: Vec<Expense>,
}

impl TransactionTableModel {
    const COLUMN_COUNT: i32 = 4;

    pub fn load_expenses(&mut self) {
        self.begin_reset_model();
        self.expenses = RUNTIME.block_on(async {
            fetch_expenses(db_pool())
                .await
                .expect("failed to load expenses")
        });
        self.end_reset_model();
    }

    pub fn add_expense(&mut self) {
        // TODO: make the default date the current date
        RUNTIME.block_on(async {
            Expense::create(Default::default(), db_pool())
                .await
                .expect("failed to create expense");
        });
        self.load_expenses();
    }
}

impl QAbstractTableModel for TransactionTableModel {
    fn row_count(&self) -> i32 {
        self.expenses.len() as i32
    }

    fn column_count(&self) -> i32 {
        Self::COLUMN_COUNT
    }

    fn data(&self, index: QModelIndex, role: i32) -> QVariant {
        if index.row() >= self.expenses.len() as i32 || index.column() >= Self::COLUMN_COUNT {
            return QVariant::default();
        }

        let expense = &self.expenses[index.row() as usize];
        const DISPLAY_ROLE: i32 = 0;
        // const EDIT_ROLE: i32 = 2;

        if role != DISPLAY_ROLE {
            return QVariant::default();
        }

        let category = expense.category.clone().unwrap_or_default().title;
        let account = expense.account.clone().unwrap_or_default().name;
        let date = expense.date.to_string();
        let amount = expense.amount.to_string();

        let variant = match index.column() {
            0 => QVariant::from(QString::from(category)),
            1 => QVariant::from(QString::from(account)),
            2 => QVariant::from(QString::from(date)),
            3 => QVariant::from(QString::from(amount)),
            _ => QVariant::default(),
        };

        variant
    }

    fn role_names(&self) -> HashMap<i32, QByteArray> {
        let mut roles = HashMap::new();
        roles.insert(0, "display".into());
        roles
    }
}

File: folio/models/transaction_table.py
from PySide6.QtCore import Qt, QAbstractListModel, QAbstractTableModel, QByteArray, Slot
from dataclasses import  dataclass
from folio.models.category_list import Category
from folio.models.account_list import Account

class Expense:
    def __init__(
        self,
        id: str = "",
        account: Account = Account(),
        category: Category = Category(),
        date: str = "",
        amount: float = 0
    ):
        self.id = id
        self.account: Account = account
        self.category: Category = category
        self.date: str = date
        self.amount: float = amount

class TransactionTableModel(QAbstractTableModel):
    COLUMN_COUNT: int = 4

    def __init__(self,parent = None):
        super().__init__(parent)
        self.expenses: list[Expense] = []

    def rowCount(self, parent = None):
        return len(self.expenses)
    
    def columnCount(self, parent = None):
        return self.COLUMN_COUNT
    
    def data(self, index,role=Qt.ItemDataRole.DisplayRole):
        if index.row() >= len(self.expenses) or index.column() >= self.columnCount():
            return None
        
        expense = self.expenses[index.row()]
        match index.column():
            case 0:
                return expense.category.title
            case 1:
                return expense.account.name
            case 2:
                return expense.date
            case 3:
                return expense.amount
        return None
    
    @Slot()
    def load_expenses(self):
        self.beginResetModel()
        self.expenses = [
            Expense(amount=0,date="2024-12-12",account=Account(name="RBC"),category=Category(title="Name")),
            Expense(amount=550,date="2024-12-12",account=Account(name="Absa"),category=Category(title="Name")),
            Expense(amount=330,date="2024-12-12",account=Account(name="FNB"),category=Category(title="Name")),
            Expense(amount=240,date="2024-12-12",account=Account(name="Zanax"),category=Category(title="Name")),
            Expense(amount=240,date="2024-12-12",account=Account(name="RBC"),category=Category(title="Name")),
            Expense(amount=120,date="2024-12-12",account=Account(name="RBC"),category=Category(title="Name")),
        ]
        self.endResetModel()
